js/modules/loki_message_api.js

```diff
@@ -32,109 +36,143 @@ class LokiServer {
+      }
 
-    const fetchOptions = {
-      method: options.type,
-      headers,
-      timeout: options.timeout,
-    };
+      const fetchOptions = {
+        method: options.type,
+        headers,
+        timeout: options.timeout,
+      };
+      let response;
+      try {
+        response = await fetch(options.url, fetchOptions);
+      } catch (e) {
+        // TODO: Maybe we shouldn't immediately delete?
+        log.error(options.type, options.url, 0, `Error retrieving messages from ${nodeUrl}`);
+        window.LokiSnodeAPI.unreachableNode(ourKey, nodeUrl);
+        throw HTTPError('fetch error', 0, e.toString());
+      }
 
-    let response;
-    try {
-      response = await fetch(options.url, fetchOptions);
-    } catch (e) {
-      log.error(options.type, options.url, 0, 'Error');
-      throw HTTPError('fetch error', 0, e.toString());
-    }
+      let result;
+      if (
+        options.responseType === 'json' &&
+        response.headers.get('Content-Type') === 'application/json'
+      ) {
+        result = await response.json();
+      } else if (options.responseType === 'arraybuffer') {
+        result = await response.buffer();
+      } else {
+        result = await response.text();
+      }
+      completedRequests += 1;
 
-    let result;
-    if (
-      options.responseType === 'json' &&
-      response.headers.get('Content-Type') === 'application/json'
-    ) {
-      result = await response.json();
-    } else if (options.responseType === 'arraybuffer') {
-      result = await response.buffer();
-    } else {
-      result = await response.text();
+      if (response.status >= 0 && response.status < 400) {
+        if (result.lastHash) {
+          window.LokiSnodeAPI.updateLastHash(nodeUrl, result.lastHash);
+        }
+        return result;
+      }
+      log.error(options.type, options.url, response.status, 'Error');
+      throw HTTPError('retrieveMessages: error response', response.status, result);
     }
 
-    if (response.status >= 0 && response.status < 400) {
-      if (result.lastHash) {
-        currentNode.lastHash = result.lastHash;
+    while (completedRequests < MINIMUM_SUCCESSFUL_REQUESTS) {
+      const remainingRequests = MINIMUM_SUCCESSFUL_REQUESTS - completedRequests;
+      ourSwarmNodes = await window.LokiSnodeAPI.getOurSwarmNodes();
+      if (Object.keys(ourSwarmNodes).length < remainingRequests) {
+        if (completedRequests !== 0) {
+          // TODO: Decide how to handle some completed requests but not enough
+        }
+        return;
       }
-      return result;
+
+      const requests = await Promise.all(
+        Object.entries(ourSwarmNodes)
+          .splice(0, remainingRequests)
+          .map(([nodeUrl, lastHash]) => doRequest(nodeUrl, lastHash).catch(() => null))
+      );
+      // Requests is now an array of null for failed requests and the json for success
+      requests.filter(v => v !== null && 'messages' in v)
+        .forEach(v => callback(v.messages));
```

Comment:
Would it be preferable to call `callback(v.message)` still as part of the promise chain?
That way we can call the callback as soon as we have valid data, without having to wait for the other requests to finish or throw.

Reply:
Or even directly in the `doRequest` function?